sn_networking/src/cmd.rs

```diff
@@ -756,13 +709,98 @@ impl SwarmDriver {
+                        if *is_bad {
+                            continue;
+                        }
+                    }
+                    self.verify_peer_quote(peer_id, quote);
+                }
+            }
         }
 
         self.log_handling(cmd_string.to_string(), start.elapsed());
 
         Ok(())
     }
 
+    fn record_node_issue(&mut self, peer_id: PeerId, issue: NodeIssue) {
+        info!("Peer {peer_id:?} is reported as having issue {issue:?}");
+        let (issue_vec, is_bad) = self.bad_nodes.entry(peer_id).or_default();
+
+        // If being considered as bad already, skip certain operations
+        if !(*is_bad) {
+            // Remove outdated entries
+            issue_vec.retain(|(_, timestamp)| timestamp.elapsed().as_secs() < 300);
+
+            // check if vec is already 10 long, if so, remove the oldest issue
+            // we only track 10 issues to avoid mem leaks
+            if issue_vec.len() == 10 {
+                issue_vec.remove(0);
+            }
+
+            // To avoid being too sensitive, only consider as a new issue
+            // when after certain while since the last one
+            let is_new_issue = if let Some((_issue, timestamp)) = issue_vec.last() {
+                timestamp.elapsed().as_secs() > 10
+            } else {
+                true
+            };
+
+            if is_new_issue {
+                issue_vec.push((issue, Instant::now()));
+            }
+
+            // Only consider candidate as a bad node when:
+            //   accumulated THREE same kind issues within certain period
+            for (issue, _timestamp) in issue_vec.iter() {
+                let issue_counts = issue_vec
+                    .iter()
+                    .filter(|(i, _timestamp)| *issue == *i)
+                    .count();
+                if issue_counts >= 3 {
+                    *is_bad = true;
+                    info!("Peer {peer_id:?} accumulated {issue_counts} times of issue {issue:?}. Consider it as a bad node now.");
+                    // Once a bad behaviour detected, no point to continue
+                    break;
+                }
+            }
+        }
+
+        if *is_bad {
+            warn!("Cleaning out bad_peer {peer_id:?}");
+            if let Some(dead_peer) = self.swarm.behaviour_mut().kademlia.remove_peer(&peer_id) {
+                self.connected_peers = self.connected_peers.saturating_sub(1);
+                self.send_event(NetworkEvent::PeerRemoved(
+                    *dead_peer.node.key.preimage(),
+                    self.connected_peers,
+                ));
+                self.log_kbuckets(&peer_id);
+                let _ = self.check_for_change_in_our_close_group();
+            }
+        }
+    }
+
+    fn verify_peer_quote(&mut self, peer_id: PeerId, quote: PaymentQuote) {
+        if let Some(history_quote) = self.quotes_history.get(&peer_id) {
+            if !history_quote.historical_verify(&quote) {
```

Comment:
you could shorten if statements like this into its own function labelled by its purpose, that way it would be much easier to gloss over if you're reading through, but no big deal 👍

Reply:
not much gain for this case actually